SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc

```diff
@@ -114,75 +122,152 @@ void HGCDigitizer::accumulate(PileUpEventPrincipal const& e, edm::EventSetup con
+      float tof(toa-dist2center/refSpeed_+tofDelay_);
+      int itime=floor( tof/bxTime_ ) ;
       
-    if(itime<0 || itime>5) continue; 
+      //no need to add bx crossing - tof comes already corrected from the mixing module
+      //itime += bxCrossing;
+      itime += 9;
       
-    //energy deposited 
-    HGCSimEn_t ien( hit_it->energy() );
+      if(itime<0 || itime>14) continue; 
+            
+      //check if already existing (perhaps could remove this in the future - 2nd event should have all defined)
+      HGCSimHitDataAccumulator::iterator simHitIt=simHitAccumulator_->find(id);
+      if(simHitIt==simHitAccumulator_->end())
+	{
+	  simHitAccumulator_->insert( std::make_pair(id,baseData) );
+	  simHitIt=simHitAccumulator_->find(id);
+	}
       
-    //check if already existing (perhaps could remove this in the future - 2nd event should have all defined)
-    HGCSimHitDataAccumulator::iterator simHitIt=simHitAccumulator_->find(id);
-    if(simHitIt==simHitAccumulator_->end()) {
-      HGCSimHitData baseData;
-      baseData.fill(0.);
-      simHitAccumulator_->insert( std::make_pair(id,baseData) );
-      simHitIt=simHitAccumulator_->find(id);
-    }
+      //check if time index is ok and store energy
+      if(itime >= (int)simHitIt->second[0].size() ) continue;
+
+      (simHitIt->second)[0][itime] += charge;
+      float accCharge=(simHitIt->second)[0][itime];
       
-    //check if time index is ok and store energy
-    if(itime >= (int)simHitIt->second.size() ) continue;
-    (simHitIt->second)[itime] += ien;
-  }
+      //time-of-arrival (check how to be used)
+      if(weightToAbyEnergy) (simHitIt->second)[1][itime] += charge*tof;
+      else if((simHitIt->second)[1][itime]==0)
+	{	
+	  if( accCharge>tdcOnset)
+	    {
+	      //extrapolate linear using previous simhit if it concerns to the same DetId
+	      float fireTDC=tof;
+	      if(i>0)
+		{
+		  uint32_t prev_id  = std::get<1>(hitRefs[i-1]);
+		  if(prev_id==id)
+		    {
+		      float prev_toa    = std::get<2>(hitRefs[i-1]);
+		      float prev_tof(prev_toa-dist2center/refSpeed_+tofDelay_);
+		      //float prev_charge = std::get<3>(hitRefs[i-1]);
+		      float deltaQ2TDCOnset = tdcOnset-((simHitIt->second)[0][itime]-charge);
+		      float deltaQ          = charge;
+		      float deltaT          = (tof-prev_tof);
+		      fireTDC               = deltaT*(deltaQ2TDCOnset/deltaQ)+prev_tof;
+		    }		  
+		}
+
+	      (simHitIt->second)[1][itime]=fireTDC;
+	    }
+	}
+    }
+  hitRefs.clear();
   
   //add base data for noise simulation
   if(!checkValidDetIds_) return;
   if(!geom.isValid()) return;
-  HGCSimHitData baseData;
-  baseData.fill(0.);
-  const std::vector<DetId> &validIds=geom->getValidDetIds(); 
+  const std::vector<DetId> &validIds=geom->getValidDetIds();   
   int nadded(0);
   if (useAllChannels_) {
     for(std::vector<DetId>::const_iterator it=validIds.begin(); it!=validIds.end(); it++) {
```

Comment:
@lgray - please, have a look at lines 275-276. Is there "else" missing? (BTW, it's a loop on ~5 million ids)